tests/tests_pytorch/test_cli.py

```diff
@@ -827,6 +828,53 @@ def configure_optimizers(self):
     assert init[1]["lr_scheduler"].gamma == 0.3
 
 
+def test_lightning_cli_load_from_checkpoint_dependency_injection(cleandir):
+    class TestModel(BoringModel):
+        def __init__(
+            self,
+            optimizer: OptimizerCallable = torch.optim.Adam,
+            scheduler: LRSchedulerCallable = torch.optim.lr_scheduler.ConstantLR,
+            activation: torch.nn.Module = lazy_instance(torch.nn.LeakyReLU, negative_slope=0.05),
+        ):
+            super().__init__()
+            self.save_hyperparameters()
+            self.optimizer = optimizer
+            self.scheduler = scheduler
+            self.activation = activation
+
+        def configure_optimizers(self):
+            optimizer = self.optimizer(self.parameters())
+            scheduler = self.scheduler(optimizer)
+            return {"optimizer": optimizer, "lr_scheduler": scheduler}
+
+    with mock.patch("sys.argv", ["any.py", "--trainer.max_epochs=1"]):
+        cli = LightningCLI(TestModel, run=False, auto_configure_optimizers=False)
+    cli.trainer.fit(cli.model)
+
+    hparams_path = Path(cli.trainer.log_dir) / "hparams.yaml"
+    assert hparams_path.is_file()
+    hparams = yaml.safe_load(hparams_path.read_text())
+    expected = {
+        "optimizer": "torch.optim.Adam",
+        "scheduler": "torch.optim.lr_scheduler.ConstantLR",
+        "activation": {"class_path": "torch.nn.LeakyReLU", "init_args": {"negative_slope": 0.05, "inplace": False}},
+    }
+    assert hparams == expected
+
+    checkpoint_path = next(Path(cli.trainer.log_dir, "checkpoints").glob("*.ckpt"), None)
+    assert checkpoint_path.is_file()
+    ckpt = torch.load(checkpoint_path)
+    assert ckpt["hyper_parameters"] == expected
+
+    model = TestModel.load_from_checkpoint(checkpoint_path, instantiator=instantiate_module)
```

Comment:
Preferably `load_from_checkpoint` should not need to be given a function to use for instantiation, but just work automatically. Though, I don't have yet a good way to do this.

Reply:
> How would this work in subclass mode where you don't have TestModel explictly?

`ModelBaseClass.load_from_checkpoint` or even `LightningModule.load_from_checkpoint`. Note that the idea is for this to be `model = Class.load_from_checkpoint(checkpoint_path)`. The instantiator should be saved in the checkpoint and be used automatically. How to do this is a point to discuss.

> Why did you choose to add an argument to LightningModule.load_from_checkpoint instead of something along the lines of cli.load_from_checkpoint()?

- There is no reason why models trained with CLI should have less features than models trained without the CLI. `load_from_checkpoint` support is missing and this PR is intended to add it. The idea is that it works the same way. No need for people to learn something new.
- `LightningCLI` is a class intended for running through the command line. Loading a model is not running from the command line, so I don't see why it should be tied to the CLI.
- `LightningCLI` is intended for reproducibility. Normally you would have `LightningCLI(..., run=True)`, or some function `def cli(): ... LightningCLI(run=False) ...` that runs something. In these cases there is no instance to do something like `cli.load_from_checkpoint()`. For this you would need to implement more stuff, which adds boilerplate.

Reply:
There hasn't been any response in a long time. I do think that `LightningModule.load_from_checkpoint` should be fixed instead of having something exclusive to cli.

> Preferably load_from_checkpoint should not need to be given a function to use for instantiation, but just work automatically. Though, I don't have yet a good way to do this.

This is pending. I guess in the checkpoint the import path to the function needed to instantiate can be saved in some key, like `instantiator="lightning.pytorch.cli.instantiate_module"`.

Also I am not sure how `save_hyperparameters` of a data module is used. Is the data module also loaded from a checkpoint? Where can I see an example of this?

Reply:
Left some comments around the context var. I have the same question as Carlos. I fear that users will forget to pass this special argument and not benefit from this feature.

Reply:
Yes, the idea is to remove the instantiator from the call. It should be just `TestModel.load_from_checkpoint(checkpoint_path)`. But it must be possible to know from the checkpoint that the instantiator from the cli must be used instead of the default one. Could be that in the checkpoint dict, there is a new key say `module_instantiator` and have as value the import path to the function to use for instantiating. This would be somewhat generic since it keeps open the possibility in the future to have a different instantiator. Instead, it could be just a boolean specific to cli instantiation, which is less generic. Another idea could be that instead of saving in the checkpoint dict, it could be in `hparams.yaml`, e.g. in the `_instantiator` key.

Reply:
Ok sounds good. I don't know which way is more complex, but it could be explored as a follow up.

One small issue I see is that right now, the instantiator is not explicitly listed in the `load_from_checkpoint` signature. It gets eaten up by kwargs and passed down. But we do some computation with the kwargs inbetween, like

`checkpoint[cls.CHECKPOINT_HYPER_PARAMS_KEY].update(kwargs)`

and so I think it would be cleaner to not include it in the varkwargs.

Reply:
I will push a commit when I have some time to work on the comments.

Reply:
I have pushed a commit. Now `load_from_checkpoint` does not require to be given an instantiator since this is saved and taken from the checkpoint.

Reply:
Now I mentioned the new feature in the docs and added a unit test for the subclass mode.